### PR TITLE
Include ignored pods when computing backoff delay for Job pod failures

### LINKED ISSUE #119378: Job: Pod backoff calculation should not depend on failure policy
### What happened?

Pod backoff calculation excludes pods that match an `Ignore` rule in the failure policy:

https://github.com/kubernetes/kubernetes/blob/847f1758748ac357e79094c5e85029a10428e65f/pkg/controller/job/job_controller.go#L1365

This could lead to fast retries, overwhelming the control plane, if most failures match an `Ignore` rule. The semantics of `Ignore` is not to count towards the limit, but the backoff should still be respected.

/sig apps

### What did you expect to happen?

Recreations to be throttled.

### How can we reproduce it (as minimally and precisely as possible)?

1. Create a Job with failure policy that matches every kind of failure.
1. Use a container that always fails

You would observe immediate recreation.

### Anything else we need to know?

_No response_

### Kubernetes version

1.27+

### Cloud provider

<details>
N/A
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by mimowo**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #119378 

#### Does this PR introduce a user-facing change?

```release-note
Fix computing backoff delay when using Job pod failure policy, by including in the backoff delay calculation pod failures ignored from the backoffLimit counter
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -951,6 +951,19 @@ func (jm *Controller) deleteActivePods(ctx context.Context, job *batch.Job, pods
 	return successfulDeletes, errorFromChannel(errCh)
 }
 
+func getNonIgnoredFailedPodsCnt(jobCtx *syncJobCtx, failedPods []*v1.Pod) int {
```
Comment (alculquicondor) [inline pkg/controller/job/job_controller.go:954]:
```suggestion
func nonIgnoredFailedPodsCount(jobCtx *syncJobCtx, failedPods []*v1.Pod) int {
```

`get` generally doesn't add any value.

**Review by alculquicondor — COMMENTED** [review]
/approve

File: pkg/controller/job/job_controller.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/job/job_controller.go]: right, cleaned up

**Comment by alculquicondor** [comment]
It could be valuable in 1.27 to protect the control plane

**Pull request merged**